### PR TITLE
[Indexing] Add advanced index casting in `_canonicalize_tuple_index`

### PR DESCRIPTION
**Pull request by makslevental**
Paves the way for advanced indexing by canonicalizing Python lists of ints to `tensor<...xindex>` values.

Stacked on https://github.com/iree-org/iree-llvm-sandbox/pull/711

### CODE REVIEW

File: test/python/dialects/indexing/dialect.py [hunk 1]
```diff
@@ -439,3 +440,37 @@ def testSimpleLiteralIndexing():
...
+@run
+def testCanonicalizeTupleIndexCastListLiteral():
+  with mlir_mod_ctx() as module:
+
+    for n_tens in range(1, 4):
+      uniqs = set()
+      n_slices = 4 - n_tens
+      ten_idx = [[0], [1]]
+      slice_idx = slice(None)
+      for p in permutations([str(ten_idx)] * n_tens +
```
Comment (ingomueller-net) [inline test/python/dialects/indexing/dialect.py:455]: A more pythonic way would be `uniqs = set(permuations(...))`. (Though you probably don't even need the `set`, do you?)

File: test/python/dialects/indexing/dialect.py [hunk 2]
```diff
@@ -439,3 +440,37 @@ def testSimpleLiteralIndexing():
     print(Tensor(w.owner.operands[1]))
     # CHECK: %{{.*}} = indexing.gather %[[TEN]][%[[CST0]]] gather_dims([0, 1, 2]) unique : (tensor<10x10x10x10xi32>, tensor<3xindex>) -> tensor<10xi32>
     print(w.owner)
+
+
+# CHECK-LABEL: TEST: testCanonicalizeTupleIndexCastListLiteral
+@run
+def testCanonicalizeTupleIndexCastListLiteral():
```
Comment (ingomueller-net) [inline test/python/dialects/indexing/dialect.py:447]: I don't understand what this test is doing. Can you explain?

File: test/python/dialects/indexing/dialect.py [hunk 3]
```diff
@@ -439,3 +440,37 @@ def testSimpleLiteralIndexing():
...
+      for p in permutations([str(ten_idx)] * n_tens +
+                            [str(slice_idx)] * n_slices):
+        uniqs.add(p)
+
+      for u in uniqs:
+        u = tuple(u)
+        tens_is = [i for i, t in enumerate(u) if t == str(ten_idx)]
+        slice_is = [i for i, s in enumerate(u) if s == str(slice_idx)]
+
+        tens_slices = _canonicalize_tuple_index(tuple(map(eval, u)), 4)
```
Comment (ingomueller-net) [inline test/python/dialects/indexing/dialect.py:464]: Can you try to find a way to do this without `eval`?

File: test/python/dialects/indexing/dialect.py (hunk 2 above)
Comment (ingomueller-net) [reply test/python/dialects/indexing/dialect.py:447]: Actually, I think that tests should be so simple that they don't need explanation. Otherwise, you'd need to test the test. Even if this gets quite a bit longer, maybe it is better to spell out the cases here?

File: python/mlir_structured/dialects/indexing.py [hunk 4]
```diff
@@ -503,8 +500,32 @@ def _is_scalar(e: Any) -> bool:
   return isinstance(e, Scalar) or isinstance(e, (int, float, bool))
 
 
-def _canonicalize_tuple_index(idx, rank: int):
-  """Helper to remove Ellipsis and replace with implicit trailing slice(None)s.
+def _is_index_tensor(x):
+  """Returns True if x is a Tensor with index dtype, False otherwise."""
+  return (isinstance(x, Value) and Tensor.isinstance(x) and
```
Comment (ingomueller-net) [inline python/mlir_structured/dialects/indexing.py:505]: Is the first test needed? Doesn't `Tensor.isinstance` subsume that test?

File: python/mlir_structured/dialects/indexing.py [hunk 5]
```diff
@@ -503,8 +500,32 @@ def _is_scalar(e: Any) -> bool:
...
+          IndexType.isinstance(x.dtype))
+
+
+def _is_int_arraylike(x):
+  """Returns True if x is array-like with integer dtype, False otherwise."""
+  return (isinstance(x, int) and not isinstance(x, bool) or
+          isinstance(x, (list, tuple)) and all(_is_int_arraylike(e) for e in x))
```
Comment (ingomueller-net) [inline python/mlir_structured/dialects/indexing.py:512]: This allows nested lists. That isn't intended, is it?

**Review by ingomueller-net — COMMENTED** [review]
I currently have trouble what the PR exactly does because I don't understand what the test does.

File: test/python/dialects/indexing/dialect.py (hunk 1 above)
Comment (makslevental) [reply test/python/dialects/indexing/dialect.py:455]: You do need set because you want to shuffle `ten_idx` around but you want the `n_tens` copies to distinguishable.

File: test/python/dialects/indexing/dialect.py (hunk 2 above)
Comment (makslevental) [reply test/python/dialects/indexing/dialect.py:447]: This test tests whether the canonicalization works for all permutations of non-trivial indexer and `:`, so e.g., 

```
ten[idx, :, :, :]
ten[:, idx, :, :]
ten[:, :, idx, :]
...
```

You get the idea. I did have lots of these cases spelled out before (I used this same code to generate the cases) but I think you can see that it's the same case just permuted. And, since this is a unit test for an unexposed method, and the only thing that needs to be verified is the `assert` (i.e., no IR prints and such), I changed it to this. I can change it back but it will really will blow up the length of this file.

File: test/python/dialects/indexing/dialect.py (hunk 3 above)
Comment (makslevental) [reply test/python/dialects/indexing/dialect.py:464]: You can't - `slice(None)` and `[[0], [1]]` objects aren't hashable (but their `str` representations are) so you wouldn't be able to find unique permutations.

File: python/mlir_structured/dialects/indexing.py (hunk 4 above)
Comment (makslevental) [reply python/mlir_structured/dialects/indexing.py:505]: The first test is needed and `Tensor.isinstance` does not subsume - if you pass something that's not an `ir.Value` you will get a pybind11 exception from failing to cast [here](https://github.com/llvm/llvm-project/blob/2713781b0cdc1af647048ec97d40101664673dee/mlir/include/mlir/Bindings/Python/PybindAdaptors.h#L507).

File: python/mlir_structured/dialects/indexing.py (hunk 5 above)
Comment (makslevental) [reply python/mlir_structured/dialects/indexing.py:512]: Why not? You should be able to pass indexers that have shapes like `2x2x2` right? Just btw, in general all of these helper methods should be sensible (I assume) because Jax aims to support numpy advanced indexing as authenticall as possible.

File: test/python/dialects/indexing/dialect.py (hunk 1 above)
Comment (ingomueller-net) [reply test/python/dialects/indexing/dialect.py:455]: Ah, right, some of the permutations will equivalent because some elements of the original list are equivalent...

**Review by nicolasvasilache — APPROVED** [review]

File: test/python/dialects/indexing/dialect.py (hunk 2 above)
Comment (nicolasvasilache) [reply test/python/dialects/indexing/dialect.py:447]: This seems fine to me, I would just add comments that you generate and test all possible versions on the fly and add 1-2 examples of MLIR that gets generated.

File: python/mlir_structured/dialects/indexing.py (hunk 5 above)
Comment (nicolasvasilache) [reply python/mlir_structured/dialects/indexing.py:512]: Generally, as I look at this PR and its dependent, I would find it helpful to have more doc comments in the form of examples that also hint at the usage, similarly to how we'd write things in MLIR ODS descriptions.

Here we are just talking about supporting underlying mechanics but I already find the polymorphism a bit hard to follow. Just listing the accepted forms in the doc would be useful to me, even if it quickly turns redundant.
